File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/interop/InteropModuleRegistryTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.bridge.interop;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import com.facebook.react.config.ReactFeatureFlags;
import com.facebook.react.modules.core.JSTimers;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import org.junit.Before;
import org.junit.Test;

public class InteropModuleRegistryTest {

  InteropModuleRegistry underTest;

  @Before
  public void setup() {
    underTest = new InteropModuleRegistry();
  }

  @Test
  public void shouldReturnInteropModule_withFabricDisabled_returnsFalse() {
    ReactFeatureFlags.enableFabricRenderer = false;

    assertFalse(underTest.shouldReturnInteropModule(RCTEventEmitter.class));
  }

  @Test
  public void shouldReturnInteropModule_withFabricInteropDisabled_returnsFalse() {
    ReactFeatureFlags.enableFabricRenderer = true;
    ReactFeatureFlags.unstable_useFabricInterop = false;

    assertFalse(underTest.shouldReturnInteropModule(RCTEventEmitter.class));
  }

  @Test
  public void shouldReturnInteropModule_withUnregisteredClass_returnsFalse() {
    ReactFeatureFlags.enableFabricRenderer = true;
    ReactFeatureFlags.unstable_useFabricInterop = true;

    assertFalse(underTest.shouldReturnInteropModule(JSTimers.class));
  }

  @Test
  public void shouldReturnInteropModule_withRegisteredClass_returnsTrue() {
    ReactFeatureFlags.enableFabricRenderer = true;
    ReactFeatureFlags.unstable_useFabricInterop = true;

    underTest.registerInteropModule(RCTEventEmitter.class, new FakeRCTEventEmitter());

    assertTrue(underTest.shouldReturnInteropModule(RCTEventEmitter.class));
  }

  @Test
  public void getInteropModule_withRegisteredClassAndInvalidFlags_returnsNull() {
    ReactFeatureFlags.enableFabricRenderer = false;
    ReactFeatureFlags.unstable_useFabricInterop = false;
    underTest.registerInteropModule(RCTEventEmitter.class, new FakeRCTEventEmitter());

    RCTEventEmitter interopModule = underTest.getInteropModule(RCTEventEmitter.class);

    assertNull(interopModule);
  }

  @Test
  public void getInteropModule_withRegisteredClassAndValidFlags_returnsInteropModule() {
    ReactFeatureFlags.enableFabricRenderer = true;
    ReactFeatureFlags.unstable_useFabricInterop = true;
    underTest.registerInteropModule(RCTEventEmitter.class, new FakeRCTEventEmitter());

    RCTEventEmitter interopModule = underTest.getInteropModule(RCTEventEmitter.class);

    assertTrue(interopModule instanceof FakeRCTEventEmitter);
  }

  @Test
  public void getInteropModule_withUnregisteredClass_returnsNull() {
    ReactFeatureFlags.enableFabricRenderer = true;
    ReactFeatureFlags.unstable_useFabricInterop = true;
    JSTimers missingModule = underTest.getInteropModule(JSTimers.class);

    assertNull(missingModule);
  }
}

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/interop/InteropModuleRegistryTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@file:Suppress(
    "DEPRECATION") // Suppressing as we want to test specifically with RCTEventEmitter here

package com.facebook.react.bridge.interop

import com.facebook.react.common.annotations.UnstableReactNativeAPI
import com.facebook.react.internal.featureflags.ReactNativeFeatureFlags
import com.facebook.react.internal.featureflags.ReactNativeFeatureFlagsDefaults
import com.facebook.react.internal.featureflags.ReactNativeFeatureFlagsForTests
import com.facebook.react.modules.core.JSTimers
import com.facebook.react.uimanager.events.RCTEventEmitter
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test

@OptIn(UnstableReactNativeAPI::class)
class InteropModuleRegistryTest {

  private lateinit var underTest: InteropModuleRegistry

  @Before
  fun setup() {
    underTest = InteropModuleRegistry()
    ReactNativeFeatureFlagsForTests.setUp()
  }

  @After
  fun tearDown() {
    ReactNativeFeatureFlags.dangerouslyReset()
  }

  @Test
  fun getInteropModule_withRegisteredClassAndInvalidFlags_returnsNull() {
    overrideFeatureFlags(false, false)
    underTest.registerInteropModule(RCTEventEmitter::class.java, FakeRCTEventEmitter())

    val interopModule = underTest.getInteropModule(RCTEventEmitter::class.java)

    assertThat(interopModule).isNull()
  }

  @Test
  fun getInteropModule_withRegisteredClassAndValidFlags_returnsInteropModule() {
    overrideFeatureFlags(true, true)
    underTest.registerInteropModule(RCTEventEmitter::class.java, FakeRCTEventEmitter())

    val interopModule = underTest.getInteropModule(RCTEventEmitter::class.java)

    assertThat(interopModule).isInstanceOf(FakeRCTEventEmitter::class.java)
  }

  @Test
  fun getInteropModule_withUnregisteredClass_returnsNull() {
    overrideFeatureFlags(true, true)
    val missingModule = underTest.getInteropModule(JSTimers::class.java)

    assertThat(missingModule).isNull()
  }

  private fun overrideFeatureFlags(useFabricInterop: Boolean, enableFabricRenderer: Boolean) {
    ReactNativeFeatureFlags.override(
        object : ReactNativeFeatureFlagsDefaults() {
          override fun useFabricInterop(): Boolean = useFabricInterop

          override fun enableFabricRenderer(): Boolean = enableFabricRenderer
        })
  }
}
